### PR TITLE
1st part of upgrading registerElement to customElement

### PR DESCRIPTION
**Pull request by mfonville**
Replace registerElement with customElement

This is a general improvement, and a prerequisite before any electron upgrade.

I was able to do it for the javascript files. But I am stuck with the coffeescript files as I have described in https://github.com/atom/atom/pull/22011#issuecomment-799753706
Those probably require some bigger overhaul that I am not able to deal with.

This should be ok for merging (maybe tests need some updates?) but if any of the devs could please help out with the other files that would be much appreciated!

### CODE REVIEW

File: spec/panel-container-element-spec.js [hunk 1]
```diff
@@ -9,7 +9,8 @@ describe('PanelContainerElement', () => {
   class TestPanelContainerItem {}
 
   class TestPanelContainerItemElement_ extends HTMLElement {
-    createdCallback() {
+    constructor()) {
```
Comment (icecream17) [inline spec/panel-container-element-spec.js:12]:
```suggestion
    constructor() {
```
Caught by the tests

Comment (mfonville) [reply spec/panel-container-element-spec.js:12]: Thanks!

**Review by aminya — CHANGES_REQUESTED** [review]
Could you rebase and also run lint?
```
./script/lint --fix
```

File: spec/panel-container-element-spec.js [hunk 2]
```diff
@@ -19,16 +20,16 @@ describe('PanelContainerElement', () => {
...
-    { prototype: TestPanelContainerItemElement_.prototype }
+    TestPanelContainerItemElement_
   );
 
   beforeEach(() => {
     jasmineContent = document.body.querySelector('#jasmine-content');
 
     atom.views.addViewProvider(TestPanelContainerItem, model =>
-      new TestPanelContainerItemElement().initialize(model)
+      new window.customElements.get('atom-test-container-item-element').initialize(model)
```
Comment (icecream17) [inline spec/panel-container-element-spec.js:32]:
```suggestion
      (new (window.customElements.get('atom-test-container-item-element'))).initialize(model)
```
Explainer:
```javascript
let customElementz = {
  getz (argument) {
    return class example {
      initialize () {return argument}
    }
  }
}
 
new customElementz.getz(15).initialize()
// Uncaught TypeError: customElementz.getz is not a constructor

(new customElementz.getz(15)).initialize()
// Uncaught TypeError: customElementz.getz is not a constructor

new (customElementz.getz(15)).initialize()
// Uncaught TypeError: customElementz.getz(...).initialize is not a constructor

(new (customElementz.getz(15))).initialize()
// 15
```

Comment (icecream17) [reply spec/panel-container-element-spec.js:32]: Actual maybe use the `document.createElement` function
```js
document.createElement('atom-test-container-item-element').initialize(model)
```

Tested with
```js
class testElement extends HTMLElement {}

customElements.define('test-element', testElement)

document.createElement('test-element').constructor // class testElement {}
document.createElement('test-element') instanceof testElement // true
```

Comment (mfonville) [reply spec/panel-container-element-spec.js:32]: Thanks!

**Comment by mfonville** [comment]
@aminya thanks for the hint. Lint caught the useless constructor. No lint errors anymore now

**Comment by icecream17** [comment]
Weirdly the tests say:
`document.createElement(...).initialize is not a function`

Trying it out on a current browser:
`Example test raised an error when creating the new element`

Found a stackoverflow: https://stackoverflow.com/questions/43836886/failed-to-construct-customelement-error-when-javascript-file-is-placed-in-head

Apparently attributes (e.g. `this.classList`) cannot be modified in a constructor.
In this case, some implementations don't enforce that, so that could explain why there's no constructor error.
But maybe it still created htmlunknownelement and that's why the tests are failing.

This affects `spec\panel-container-element-spec.js`, `src\pane-container-element.js`, `src\pane-element`, `src\panel-container-element.js`, `src\text-container-element.js`, and `src\workspace-element.js`

Maybe there should be a `constructor2` method, and then a function `createCustomElement` that automatically calls both

____
Also there's another thing about createElement

```js
class example1 extends HTMLElement {}
customElements.define('example-one', example1)
document.createElement('example-one')

// But here....
class example2 extends HTMLDivElement {}
customElements.define('example-two', example2, { extends: 'div' })
document.createElement('div', { is: 'example-two' }) // different
```

So ~line 1352 in `spec/text-editor-element-spec.js` should be changed
```js
      const parent = document.createElement(
---     'text-editor-component-test-element'
+++     'div', { is: 'text-editor-component-test-element' }
      );
```

**Comment by mfonville** [comment]
Thanks for all your feedback and research @icecream17 

Maybe one of the official Atom developers can propose what they think is the preferred approach forward? @sadick254 @darangi

**Comment by denisdeoniss** [comment]
- [ ]

**Comment by elsaxo** [comment]
> Replace registerElement with customElement
> 
> This is a general improvement, and a prerequisite before any electron upgrade.
> 
> I was able to do it for the javascript files. But I am stuck with the coffeescript files as I have described in [#22011 (comment)](https://github.com/atom/atom/pull/22011#issuecomment-799753706)
> Those probably require some bigger overhaul that I am not able to deal with.
> 
> This should be ok for merging (maybe tests need some updates?) but if any of the devs could please help out with the other files that would be much appreciated!

**Comment by icecream17** [comment]
Idea!

```js
function customDefine (name, constructor, options) {
   customElements.define(name, constructor, options)

   if (options.extends) {
      constructor.create = function createElement () {
         return document.createElement(options.extends, { is: name })
      }
   } else {
      constructor.create = function createElement () {
         return document.createElement(name)
      }
   }
}
```

**Comment by icecream17** [comment]
Also, I just realized `customElements.get` !== `document.createElement`, so maybe undo my suggestion above

**Comment by aminya** [comment]
We recently switched from `registerElement` to a `customElement` in `atomic-terminal`. The PR might be of your interest:
https://github.com/atom-community/terminal/pull/70

**Comment by mfonville** [comment]
Thanks @aminya 

Maybe you want to update this PR accordingly? (I am sorry to say that I won't be able to pursue and further improve the PR myself)

**Comment by aminya** [comment]
> Thanks @aminya
> 
> Maybe you want to update this PR accordingly? (I am sorry to say that I won't be able to pursue and further improve the PR myself)

Yes, I can take a look in the following weeks. 

> @icecream17  Maybe there should be a constructor2 method, and then a function createCustomElement that automatically calls both? (Trying to think of a better solution...)

I don't see a point in having a separate `initialize` method. This part of the code probably comes from the days of CoffeeScript. The code of `initialize` should be merged into `constructor`.

**Comment by mfonville** [comment]
This PR has been surpassed by @sadick254's new patches

**Pull request closed without merging**